spmd/compiler/fusion.py

```diff
@@ -0,0 +1,182 @@
+                self.first = node
+                break
+
+        self.output = get_output_node(gm)
+        assert (
+            self.output is not None
+        ), f"unable to locate output node in gm {gm.graph}"
+
+        if debug:
+            print(
+                f"updated info - len = {self.len} input = {self.first}, output = {self.output}"
+            )
+
+
+def _insert_buffer_node(
+    gm: fx.GraphModule, insert_before_node: fx.Node, buffer_size: Iterable[int]
+) -> fx.Node:
+    """insert a torch.empty node in front of insert_before_node"""
+    with gm.graph.inserting_before(insert_before_node):
+        new_buffer_node = gm.graph.create_node(
+            OP.CALL_FUNCTION,
+            target=torch.empty,
+            args=tuple(buffer_size),  # TODO - need device
+        )
+    assert (
+        new_buffer_node is not None
+    ), f"failed to create buffer node, size={buffer_size}"
+
+    return new_buffer_node
+
+
+def _scan_graph_for_fusion_elements(
+    gm: fx.GraphModule,
+    comm_type: Comm_Type = Comm_Type.allreduce,
+) -> Optional[List[FusionElement]]:
+    """scan entire graph for matching sections of CommTensor style expansions
+    returns list of FusionElements that match comm_type"""
+
+    element_list = []
+
+    fe_pattern = [
+        "clone",
+        "_tensor_constant",
+        "_tensor_constant",
+        comm_type,
+        "comm_result",
+        "getitem",
+        "getitem",
+        "wait_comm",
+    ]
+
+    fe_size = len(fe_pattern) - 1
+    curr_count = 0
+    curr_node_list = []
+    rank = dist.get_rank()
+
+    for i, node in enumerate(gm.graph.nodes):
+
+        pattern = fe_pattern[curr_count]
+
+        if curr_count < fe_size:
+
+            if node.name.startswith(pattern):
+                curr_node_list.append(node)
+                curr_count += 1
+                continue
+            else:
+                curr_count == 0
+                curr_node_list.clear()
+                continue
+
+        if curr_count == fe_size:
+            # should be last node
+            if node.name.startswith(pattern):
+                curr_node_list.append(node)
+
+                fe = FusionElement(
+                    comm_type=comm_type, node_list=deepcopy(curr_node_list)
+                )
+
+                # need to fully populate this fe...
+                # we will be removing/rewriting the node list so we save prev and next
+                fe.prev_node = fe.node_list[0].prev
+                fe.next_node = node.next
+
+                fe.output_name = node.name
+
+                fe.clone_node = curr_node_list[0]  # fe.node_list[0]
+                fe.comm_node = fe.node_list[3]
+                fe.wait_node = node
+
+                # compute size of this fe
+                fe.size = get_node_tensor_numel(fe.clone_node)  # type: ignore
+                element_list.append(fe)
+
+            curr_count = 0
+            curr_node_list.clear()
+            continue
```

Comment:
This continue is not needed.

Reply:
You are correct.  I'd like to keep it though for clarity and a tiny bit for efficiency.  (i.e. why fall down and eval various if's below it that won't be fired, when we know we want to continue here).